### PR TITLE
feat: add support for `Stream: true` to completionEndpoint

### LINKED ISSUE #416: `completionEndpoint` should support `"stream": true`
this is currently not implemented, so is not following the OpenAI spec

investigation and discovery in the discord [here](https://discord.com/channels/1096914990004457512/1111727007299014696), but the crux of the matter is that `completionEndpoint` in `api/openai.go` simply doesn't support streaming

discord thread copy/pasted for accessibility

<details><summary>Details</summary>
<p>

> athousandcups: I'm using my model that is `llama.cpp` based, and sending `"stream": true`, but it's not streaming - you can see that it receives `"stream": true` but then sets `Stream:false` in the `Parameter Config`:

```
1:05AM DBG Request received: {"model":"ggml-model-q4_0.bin","file":"","language":"","response_format":"","size":"","prompt":"a long time ago in a galaxy far, far away","instruction":"","input":null,"stop":null,"messages":null,"stream":true,"echo":false,"top_p":0,"top_k":0,"temperature":0,"max_tokens":632,"n":0,"batch":0,"f16":false,"ignore_eos":false,"repeat_penalty":0,"n_keep":0,"mirostat_eta":0,"mirostat_tau":0,"mirostat":0,"seed":0,"mode":0,"step":0}
1:05AM DBG Parameter Config: &{OpenAIRequest:{Model:ggml-model-q4_0.bin File: Language: ResponseFormat: Size: Prompt:<nil> Instruction: Input:<nil> Stop:<nil> Messages:[] Stream:false Echo:false TopP:0.7 TopK:80 Temperature:0.9 Maxtokens:632 N:0 Batch:0 F16:false IgnoreEOS:false RepeatPenalty:0 Keep:0 MirostatETA:0 MirostatTAU:0 Mirostat:0 Seed:0 Mode:0 Step:0} Name: StopWords:[] Cutstrings:[] TrimSpace:[] ContextSize:512 F16:false Threads:4 Debug:true Roles:map[] Embeddings:false Backend: TemplateConfig:{Completion: Chat: Edit:} MirostatETA:0 MirostatTAU:0 Mirostat:0 NGPULayers:0 ImageGenerationAssets: PromptStrings:[a long time ago in a galaxy far, far away] InputStrings:[] InputToken:[]}
```
> mudler: can you paste more logs? do you see a "Stream request received" message in the debug calls? token stream works here (see the chatbot-ui example)

> athousandcups: I don't see that - here's the full relevant log:

```[172.17.0.1]:44764  200  -  POST     /v1/completions
1:05AM DBG Request received: {"model":"ggml-model-q4_0.bin","file":"","language":"","response_format":"","size":"","prompt":"a long time ago in a galaxy far, far away","instruction":"","input":null,"stop":null,"messages":null,"stream":true,"echo":false,"top_p":0,"top_k":0,"temperature":0,"max_tokens":632,"n":0,"batch":0,"f16":false,"ignore_eos":false,"repeat_penalty":0,"n_keep":0,"mirostat_eta":0,"mirostat_tau":0,"mirostat":0,"seed":0,"mode":0,"step":0}
1:05AM DBG Parameter Config: &{OpenAIRequest:{Model:ggml-model-q4_0.bin File: Language: ResponseFormat: Size: Prompt:<nil> Instruction: Input:<nil> Stop:<nil> Messages:[] Stream:false Echo:false TopP:0.7 TopK:80 Temperature:0.9 Maxtokens:632 N:0 Batch:0 F16:false IgnoreEOS:false RepeatPenalty:0 Keep:0 MirostatETA:0 MirostatTAU:0 Mirostat:0 Seed:0 Mode:0 Step:0} Name: StopWords:[] Cutstrings:[] TrimSpace:[] ContextSize:512 F16:false Threads:4 Debug:true Roles:map[] Embeddings:false Backend: TemplateConfig:{Completion: Chat: Edit:} MirostatETA:0 MirostatTAU:0 Mirostat:0 NGPULayers:0 ImageGenerationAssets: PromptStrings:[a long time ago in a galaxy far, far away] InputStrings:[] InputToken:[]}
1:05AM DBG Loading model 'ggml-model-q4_0.bin' greedly
1:05AM DBG Model 'ggml-model-q4_0.bin' already loaded

llama_print_timings:        load time = 27870.98 ms
llama_print_timings:      sample time =   440.84 ms /   632 runs   (    0.70 ms per token)
llama_print_timings: prompt eval time = 59073.37 ms /   263 tokens (  224.61 ms per token)
llama_print_timings:        eval time = 146611.34 ms /   630 runs   (  232.72 ms per token)
llama_print_timings:       total time = 3903657.38 ms
1:08AM DBG Response: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" | but actually just yesterday... | in my head at least | I'm not really sure how to a
```
> looking at the code, in `completionInput` we first call `readInput`, which correctly prints the `Request received:` as having `"stream":true`
> so somewhere between there and the `Parameter Config` debug `Stream` either doesn't get parsed or gets reset to `false`
> ok I added a log `log.Debug().Msgf("`input`: %+v", input)` and it shows `Stream:true`
> it looks like what's happening is `readConfig` isn't finding a config file, so it's creating a new `&Config` with `OpenAIRequest: defaultRequest(modelFile)`, and `defaultRequest` seems to instantiate `Stream:false`
> and then `updateConfig` doesn't seem check if `input.Stream` is `true`
> oh wait! I thikn the problem here is that I'm using the `completionEndpoint`, but `input.Stream` is only checked in the `chatEndpoint` - so streaming is only supported in the `chatEndpoint` ?

> mudler: yes, correct, please open up a new issue, streaming should be supported also on completionEndpoint too

</p>
</details>

### PR DESCRIPTION
**Pull request by samm81**
**Description**

This PR fixes #416 

**Notes for Reviewers**
sorry for stepping on your toes @krishnaduttPanchagnula , ([pr here](https://github.com/go-skynet/LocalAI/pull/425/files)) but I wanted to get it working for myself, so took a stab at it, then when I figured it out I supposed I should commit it back :)

**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [X] Yes, I signed my commits.

### CODE REVIEW

**Comment by samm81** [comment]
working on implementing a test, but wanted to get this out first

**Comment by samm81** [comment]
I don't know why the commit is showing up as `unverified` :thinking: I don't see where in the github settings to tell github that I used my ssh key...

**Comment by samm81** [comment]
> I don't know why the commit is showing up as `unverified` thinking I don't see where in the github settings to tell github that I used my ssh key...

figured it out

File: api/openai.go [hunk 1]
```diff
@@ -143,55 +144,111 @@ func defaultRequest(modelFile string) OpenAIRequest {
...
-		for _, i := range config.PromptStrings {
-			// A model can have a "file.bin.tmpl" file associated with a prompt template prefix
-			templatedInput, err := o.loader.TemplatePrefix(templateFile, struct {
-				Input string
-			}{Input: i})
-			if err == nil {
-				i = templatedInput
-				log.Debug().Msgf("Template found, input modified to: %s", i)
-			}
+		if (len(config.PromptStrings) > 1) {
```
Comment (mudler) [inline api/openai.go:193]: is this really necessary also when stream is false? I think it's safer to keep the old behavior around and enforce this only when stream = true

**Comment by mudler** [comment]
Fantastic, thanks @samm81!  I think we are almost there :+1:

File: api/openai.go (hunk 1 above)
Comment (krishnaduttPanchagnula) [reply api/openai.go:193]: @mudler ya i agree with keeping the old behaviour

Comment (samm81) [reply api/openai.go:193]: yes meant to leave a comment - I don't have access to the openai api so I don't know what is supposed to happen here. I didn't know how to handle the multiple prompt string case so I put this in here

agreed it should only be in the `Stream: true` case, will change it

do you know what the behavior _should_ be in the `Stream: true` case? does anyone have access to the openai api that could run a query against it?

**Comment by samm81** [comment]
updated, ptal!

File: api/openai.go [hunk 2]
```diff
@@ -143,55 +144,111 @@ func defaultRequest(modelFile string) OpenAIRequest {
...
+		}
+
 		templateFile := config.Model
 
 		if config.TemplateConfig.Completion != "" {
 			templateFile = config.TemplateConfig.Completion
 		}
 
-		var result []Choice
-		for _, i := range config.PromptStrings {
```
Comment (mudler) [inline api/openai.go:167]: just keep this for loop, and make another block before implementing the input.Stream logic that looks only at the first

File: api/openai.go [hunk 3]
```diff
@@ -143,55 +144,111 @@ func defaultRequest(modelFile string) OpenAIRequest {
...
-			// A model can have a "file.bin.tmpl" file associated with a prompt template prefix
-			templatedInput, err := o.loader.TemplatePrefix(templateFile, struct {
-				Input string
-			}{Input: i})
-			if err == nil {
-				i = templatedInput
-				log.Debug().Msgf("Template found, input modified to: %s", i)
-			}
+		predInput := config.PromptStrings[0]
```
Comment (mudler) [inline api/openai.go:193]: in this way we always template the first prompt nevertheless

File: api/openai.go (hunk 2 above)
Comment (samm81) [reply api/openai.go:167]: fixed

File: api/openai.go (hunk 3 above)
Comment (samm81) [reply api/openai.go]: fixed

**Comment by samm81** [comment]
should be good to go!

**Comment by samm81** [comment]
testing:

<details><summary>Details</summary>

```
$ curl -N http://localhost:8080/v1/completions -H "Content-Type: application/json" -d '{
     "model": "ggml-model-q4_0.bin",
     "prompt": "a long time ago in a galaxy far, far away",
     "max_tokens": 32
   }'
{"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" | we made it to the Chinese Garden! | and I was so happy 😁 | I wish they had more food options but still it'"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}%
$ curl -N http://localhost:8080/v1/completions -H "Content-Type: application/json" -d '{
     "model": "ggml-model-q4_0.bin",
     "prompt": ["a long time ago in a galaxy far, far away", "went to a crazy party | "],
     "max_tokens": 32,
     "temperature": 0.7
   }'
{"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" | I was getting ready to go to bed | and Sharon came upstairs | she said something about the carpet that was on the floor of our"},{"text":"爽亮饭店 | met some people | but didn't feel like talking to them much, they were all too loud and"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}%
$ curl -N http://localhost:8080/v1/completions -H "Content-Type: application/json" -d '{
     "model": "ggml-model-q4_0.bin",
     "prompt": ["a long time ago in a galaxy far, far away", "went to a crazy party"],
     "max_tokens": 32,
     "temperature": 0.7,
     "stream": true
   }'
{"error":{"code":500,"message":"cannot handle more than 1 `PromptStrings` when `Stream`ing","type":""}}%
$ curl -N http://localhost:8080/v1/completions -H "Content-Type: application/json" -d '{
     "model": "ggml-model-q4_0.bin",
     "prompt": "a long time ago in a galaxy far, far away",
     "max_tokens": 32,
     "temperature": 0.7,
     "stream": true
   }'
data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" |"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" I"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" found"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" my"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" way"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" to"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" the"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" train"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" station"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":"!"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" |"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" But"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" it"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":"'"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":"s"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" really"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" small"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":"..."}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" |"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" On"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" the"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" train"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" now"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" "}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":"\ufffd"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":"\ufffd"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":"\ufffd"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":"\ufffd"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" |"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" Train"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" is"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"object":"text_completion","model":"ggml-model-q4_0.bin","choices":[{"text":" full"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: {"model":"ggml-model-q4_0.bin","choices":[{"finish_reason":"stop"}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}

data: [DONE]
```

</details>

**Review by mudler — APPROVED** [review]
Looking good! thanks @samm81 !

**Pull request merged**